manager/raftpicker/raftpicker.go

```diff
@@ -4,164 +4,115 @@ import (
 	"sync"
 	"time"
 
-	"golang.org/x/net/context"
+	"github.com/Sirupsen/logrus"
+
 	"google.golang.org/grpc"
-	"google.golang.org/grpc/transport"
 )
 
-// picker always picks address of cluster leader.
-type picker struct {
-	mu   sync.Mutex
+// ConnSelector is struct for obtaining connection connected to cluster leader.
+type ConnSelector struct {
+	mu      sync.Mutex
+	cluster RaftCluster
+	opts    []grpc.DialOption
+
+	cc   *grpc.ClientConn
 	addr string
-	raft AddrSelector
-	conn *grpc.Conn
 
 	stop chan struct{}
-	done chan struct{}
 }
 
-func newPicker(raft AddrSelector, addr string) *picker {
-	return &picker{
-		raft: raft,
-		addr: addr,
-
-		stop: make(chan struct{}),
-		done: make(chan struct{}),
+// NewConnSelector returns new ConnSelector with cluster and grpc.DialOpts which
+// will be used for connection create.
+func NewConnSelector(cluster RaftCluster, opts ...grpc.DialOption) *ConnSelector {
+	cs := &ConnSelector{
+		cluster: cluster,
+		opts:    opts,
+		stop:    make(chan struct{}),
 	}
+	go cs.updateLoop()
+	return cs
 }
 
-// Init does initial processing for the Picker, e.g., initiate some connections.
-func (p *picker) Init(cc *grpc.ClientConn) error {
-	conn, err := grpc.NewConn(cc)
+// Conn returns *grpc.ClientConn which connected to cluster leader.
+// It can return error if cluster wasn't ready at the moment of initial call.
+func (c *ConnSelector) Conn() (*grpc.ClientConn, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.cc != nil {
+		return c.cc, nil
+	}
+	addr, err := c.cluster.LeaderAddr()
 	if err != nil {
-		return err
+		return nil, err
 	}
-	p.conn = conn
-	return nil
-}
-
-// Pick blocks until either a transport.ClientTransport is ready for the upcoming RPC
-// or some error happens.
-func (p *picker) Pick(ctx context.Context) (transport.ClientTransport, error) {
-	if err := p.updateConn(); err != nil {
+	cc, err := grpc.Dial(addr, c.opts...)
+	if err != nil {
 		return nil, err
 	}
-	return p.conn.Wait(ctx)
+	c.cc = cc
+	c.addr = addr
+	return cc, nil
 }
 
-// PickAddr picks a peer address for connecting. This will be called repeated for
-// connecting/reconnecting.
-func (p *picker) PickAddr() (string, error) {
-	addr, err := p.raft.LeaderAddr()
+// Reset recreates underlying connection.
+func (c *ConnSelector) Reset() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.cc != nil {
+		c.cc.Close()
```

Comment:
Is double `Close()` allowed or not? I think there may be paths leading to double close. If it's a problem, it may make sense to set `c.cc = nil` right after `c.cc.Close()`.